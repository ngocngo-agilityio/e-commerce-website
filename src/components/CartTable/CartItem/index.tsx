'use client';

// Libs
import { memo, useCallback, useMemo, useState } from 'react';
import { Box, Center, Flex, Hide, Show, Td, Tr, Text } from '@chakra-ui/react';
import Image from 'next/image';
import Link from 'next/link';
import dynamic from 'next/dynamic';

// Utils
import { formatCurrency } from '@utils';

// Constants
import { APP_ROUTERS } from '@constants';

// Components
import { Counter } from '@components';
const ConfirmModal = dynamic(() => import('../../ConfirmModal'));

interface Props {
  id: number;
  name: string;
  image: string;
  price: number;
  quantity: number;
  onRemoveProduct: (id: number) => void;
  onQuantityChange: (id: number, quantity: number) => void;
}

const CartItem = ({
  id,
  name,
  image,
  price = 0,
  quantity = 1,
  onRemoveProduct,
  onQuantityChange,
}: Props) => {
  const [isOpen, setIsOpen] = useState(false);

  const total = useMemo(() => price * quantity, [price, quantity]);
  const formattedPrice = useMemo(() => formatCurrency(price), [price]);
  const formattedTotal = useMemo(() => formatCurrency(total), [total]);

  const handleOpenConfirmModal = useCallback(() => {
    setIsOpen(true);
  }, []);

  const handleCancel = useCallback(() => {
    setIsOpen(false);
  }, []);

  const handleRemoveCartItem = useCallback(() => {
    onRemoveProduct(id);
    setIsOpen(false);
  }, [id, onRemoveProduct]);

  const handleQuantityChange = useCallback(
    (quantity: number) => {
      onQuantityChange(id, quantity);
    },
    [id, onQuantityChange],
  );

  return (
    <>
      <Tr>
        <Td pl={0}>
          <Flex gap={5} alignContent="center">
            <Center
              textAlign="left"
              cursor="pointer"
              w="fit-content"
              color="removeProduct"
              onClick={handleOpenConfirmModal}
            >
              x
            </Center>

            <Hide above="md">
              <Box w={20} h={20} pos="relative">
                <Image
                  src={image}
                  alt="Product Image"
                  fill
                  objectFit="cover"
                  sizes="(max-width: 768px) 25vw, 22vw"
                  priority
                />
              </Box>

              <Flex flexDir="column" gap={2} w="full">
                <Link href={APP_ROUTERS.PRODUCT_DETAIL_PAGE(id?.toString())}>
                  <Text fontFamily="baloo" style={{ fontWeight: 'bold' }}>
                    {name}
                  </Text>
                </Link>

                <Flex
                  alignItems="center"
                  justifyContent="space-between"
                  w="full"
                >
                  <Text> {formattedPrice}</Text>

                  <Counter
                    initialQuantity={quantity}
                    onQuantityChange={handleQuantityChange}
                  />
                </Flex>

                <Text color="cart.totalAmount">Total: {formattedTotal}</Text>
              </Flex>
            </Hide>
          </Flex>
        </Td>

        <Show above="md">
          <Td>
            <Box w={12} h={14} pos="relative">
              <Image
                src={image}
                alt="Product Image"
                fill
                objectFit="cover"
                sizes="(max-width: 768px) 25vw, 22vw"
                priority
              />
            </Box>
          </Td>
          <Td>
            <Link href={APP_ROUTERS.PRODUCT_DETAIL_PAGE(id?.toString())}>
              <Text fontFamily="baloo" style={{ fontWeight: 'bold' }}>
                {name}
              </Text>
            </Link>
          </Td>
          <Td>{formattedPrice}</Td>
          <Td>
            <Counter
              initialQuantity={quantity}
              onQuantityChange={handleQuantityChange}
            />
          </Td>
          <Td color="cart.totalAmount">{formattedTotal}</Td>
        </Show>
      </Tr>
      {isOpen && (
        <ConfirmModal
          title="Delete"
          textCancel="No"
          textSubmit="Yes"
          message="Are you sure remove this product from your cart?"
          onSubmit={handleRemoveCartItem}
          onClose={handleCancel}
        />
      )}
    </>
  );
};

export default memo(CartItem);
