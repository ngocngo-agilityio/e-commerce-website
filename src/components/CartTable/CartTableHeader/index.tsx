// Libs
import { memo } from 'react';
import { Th, Thead, Tr } from '@chakra-ui/react';

const CartTableHeader = () => {
  return (
    <Thead>
      <Tr>
        <Th></Th>
        <Th></Th>
        <Th>Product</Th>
        <Th>Price</Th>
        <Th>Quantity</Th>
        <Th>Total</Th>
      </Tr>
    </Thead>
  );
};

export default memo(CartTableHeader);
