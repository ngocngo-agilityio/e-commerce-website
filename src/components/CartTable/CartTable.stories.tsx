import type { Meta, StoryObj } from '@storybook/react';

// Components
import CartTable from '.';

// Mocks
import { MOCK_CART_ITEMS } from '@mocks';

const meta = {
  title: 'CartTable',
  component: CartTable,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof CartTable>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    cart: MOCK_CART_ITEMS,
    onCartItemRemove: () => {},
    onQuantityChange: () => {},
  },
};
