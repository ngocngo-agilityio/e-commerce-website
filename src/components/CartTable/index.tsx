'use client';

// Libs
import { memo } from 'react';
import isEqual from 'react-fast-compare';
import { Show, Table, Tbody, Td, Tr } from '@chakra-ui/react';

// Types
import { CartItem as ICartItem } from '@types';

// Components
import CartTableHeader from './CartTableHeader';
import CartItem from './CartItem';

interface Props {
  cart: ICartItem[];
  onCartItemRemove: (cartId: number) => void;
  onQuantityChange: (cartId: number, quantity: number) => void;
}

const CartTable = ({
  cart = [],
  onCartItemRemove,
  onQuantityChange,
}: Props): JSX.Element => {
  return (
    <Table>
      <Show above="md">
        <CartTableHeader />
      </Show>

      <Tbody>
        {cart.length > 0 ? (
          cart.map((item) => {
            const {
              id,
              name = '',
              image = '',
              price = 0,
              quantity = 0,
            } = item || {};
            return (
              <CartItem
                key={id}
                id={id}
                name={name}
                image={image}
                price={price}
                quantity={quantity}
                onRemoveProduct={onCartItemRemove}
                onQuantityChange={onQuantityChange}
              />
            );
          })
        ) : (
          <Tr>
            <Td colSpan={6} textAlign="center" borderBottom="none">
              No products in your cart.
            </Td>
          </Tr>
        )}
      </Tbody>
    </Table>
  );
};

export default memo(CartTable, isEqual);
