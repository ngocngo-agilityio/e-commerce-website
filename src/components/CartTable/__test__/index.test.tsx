// Libs
import { render, screen } from '@testing-library/react';

// Types
import { CartItem as ICartItem } from '@types';

// Components
import CartTable from '..';

// Mocks
import { MOCK_CART_ITEMS } from '@mocks';

const mockProps = {
  cart: MOCK_CART_ITEMS,
  onQuantityChange: jest.fn(),
  onCartItemRemove: jest.fn(),
};

describe('CartTable component', () => {
  test('should match snapshot for CartTable component', () => {
    const { container } = render(<CartTable {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  test('renders no products message in cart when data is empty', () => {
    render(<CartTable {...mockProps} cart={[]} />);

    expect(screen.getByText('No products in your cart.')).toBeInTheDocument();
  });

  test('should render CartTable component successfully when has null item', () => {
    const { container } = render(
      <CartTable {...mockProps} cart={[null] as unknown as ICartItem[]} />,
    );

    expect(container).toBeInTheDocument();
  });
});
