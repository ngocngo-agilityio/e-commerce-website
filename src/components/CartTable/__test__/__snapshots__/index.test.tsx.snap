// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartTable component should match snapshot for CartTable component 1`] = `
<div>
  <table
    class="chakra-table css-0"
  >
    <thead
      class="css-0"
    >
      <tr
        class="css-0"
      >
        <th
          class="css-0"
        />
        <th
          class="css-0"
        />
        <th
          class="css-0"
        >
          Product
        </th>
        <th
          class="css-0"
        >
          Price
        </th>
        <th
          class="css-0"
        >
          Quantity
        </th>
        <th
          class="css-0"
        >
          Total
        </th>
      </tr>
    </thead>
    <tbody
      class="css-0"
    >
      <tr
        class="css-0"
      >
        <td
          class="css-1a88wtd"
        >
          <div
            class="css-ie49k9"
          >
            x
          </div>
        </td>
        <td
          class="css-0"
        >
          <img
            alt="Product Image"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="56"
            src="/_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=96&q=75"
            srcset="/_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=48&q=75 1x, /_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=96&q=75 2x"
            style="color: transparent;"
            width="48"
          />
        </td>
        <td
          class="css-0"
          style="font-weight: bold;"
        >
          <a
            href="/products/1"
          >
            Plain White Shirt
          </a>
        </td>
        <td
          class="css-0"
        >
          $69.00
        </td>
        <td
          class="css-0"
        >
          <div
            class="css-wpwawf"
          >
            <div
              class="css-1en42kf"
            >
              <button
                class="chakra-button css-15ey06b"
                type="button"
              >
                -
              </button>
            </div>
            <div
              class="css-1en42kf"
            >
              <span
                class="chakra-text css-uh9d7y"
              >
                2
              </span>
            </div>
            <div
              class="css-1en42kf"
            >
              <button
                class="chakra-button css-d83nk5"
                type="button"
              >
                +
              </button>
            </div>
          </div>
        </td>
        <td
          class="css-0"
        >
          $138.00
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
